ash-window/src/lib.rs

```diff
@@ -9,118 +9,182 @@ use ash::{
-                )
-                .window(window.window.get());
-            let surface_fn = xcb_surface::Instance::load(entry, instance);
-            surface_fn.create_xcb_surface(&surface_desc, allocation_callbacks)
-        }
-
-        (RawDisplayHandle::Android(_), RawWindowHandle::AndroidNdk(window)) => {
-            let surface_desc =
-                vk::AndroidSurfaceCreateInfoKHR::default().window(window.a_native_window.as_ptr());
-            let surface_fn = android_surface::Instance::load(entry, instance);
-            surface_fn.create_android_surface(&surface_desc, allocation_callbacks)
-        }
-
-        #[cfg(target_os = "macos")]
-        (RawDisplayHandle::AppKit(_), RawWindowHandle::AppKit(window)) => {
-            use raw_window_metal::{appkit, Layer};
-
-            let layer = match appkit::metal_layer_from_handle(window) {
-                Layer::Existing(layer) | Layer::Allocated(layer) => layer.cast(),
-            };
-
-            let surface_desc = vk::MetalSurfaceCreateInfoEXT::default().layer(&*layer);
-            let surface_fn = metal_surface::Instance::load(entry, instance);
-            surface_fn.create_metal_surface(&surface_desc, allocation_callbacks)
-        }
-
-        #[cfg(target_os = "ios")]
-        (RawDisplayHandle::UiKit(_), RawWindowHandle::UiKit(window)) => {
-            use raw_window_metal::{uikit, Layer};
+#[derive(Clone)]
+enum SurfaceExtension {
+    Windows(win32_surface::Instance),
+    Wayland(
+        raw_window_handle::WaylandDisplayHandle,
+        wayland_surface::Instance,
+    ),
+    Xlib(raw_window_handle::XlibDisplayHandle, xlib_surface::Instance),
+    Xcb(raw_window_handle::XcbDisplayHandle, xcb_surface::Instance),
+    Android(android_surface::Instance),
+    #[cfg(target_os = "macos")]
+    AppKit(metal_surface::Instance),
+    #[cfg(target_os = "ios")]
+    UiKit(metal_surface::Instance),
+}
 
-            let layer = match uikit::metal_layer_from_handle(window) {
-                Layer::Existing(layer) | Layer::Allocated(layer) => layer.cast(),
-            };
+/// Holder for a loaded platform-specific Vulkan `Surface` extension, used to create surfaces.
+///
+/// Also stores the platform-specific [`raw_window_handle::RawDisplayHandle`] variant if necessary
+/// to create surfaces, identifying the selected display server handle that was used to load the
+/// relevant extension when creating a [`SurfaceFactory`].
+#[derive(Clone)]
+pub struct SurfaceFactory(SurfaceExtension);
+
+impl SurfaceFactory {
+    /// Load the relevant surface extension for a given [`RawDisplayHandle`].
+    ///
+    // TODO: Safety docs about Entry/Instance lifetime?
+    ///
+    /// `instance` must have been created with platform specific surface extensions enabled, acquired
+    /// through [`enumerate_required_extensions()`].
+    pub fn new(
+        entry: &Entry,
+        instance: &Instance,
+        // TODO: Return "lifetimed" on DisplayHandle?
```

Comment:
I like this idea, but it might give users a false sense of security if we don't also track the lifetime of the window handle, which would require bundling a lifetime with the `SurfaceKHR` somehow. I'd leave this for future work -- updating `ash-window` should be much less disruptive than updating `ash`, anyway.